### PR TITLE
Adds validation methods to the driver

### PR DESCRIPTION
**Pull request by elijahbenizzy**
## Changes
- `validate_materialization`
- `validate_execution`

## How I tested this
- Unit tests

## Checklist

- [x] PR has an informative and human-readable title (this will be pulled into the release notes)
- [ ] Changes are limited to a single goal (no scope creep)
- [x] Code passed the pre-commit check & code is left cleaner/nicer than when first encountered.
- [x] Any _change_ in functionality is tested
- [x] New functions are documented (with a description, list of inputs, and expected output)
- [x] Placeholder code is flagged / future TODOs are captured in comments
- [x] Project documentation has been updated if adding/changing functionality.

### CODE REVIEW

**Comment by elijahbenizzy** [comment]
See #484 for more motivation

File: hamilton/driver.py [hunk 1]
```diff
@@ -1062,6 +1062,56 @@ def visualize_materialization(
...
+        inputs: Dict[str, Any] = None,
+    ):
+        """Validates execution of the graph. One can call this to validate execution, independently of actually executing.
+        Note this has no return -- it will raise a ValueError if there is an issue.
+
+        :param fg:
+        :param final_vars:
+        :param overrides:
+        :param inputs:
+        :return:
```
Comment (skrawcz) [inline hamilton/driver.py:1078]: change to :raises ValueError: if there's an error.

File: hamilton/driver.py [hunk 2]
```diff
@@ -1062,6 +1062,56 @@ def visualize_materialization(
...
+        inputs: Dict[str, Any] = None,
+    ):
+        """Validates materialization of the graph. Effectively .materialize() with a dry-run.
+        Note this has no return -- it will raise a ValueError if there is an issue.
+
+        :param materializers: Materializers to use, see the materialize() function
+        :param additional_vars: Additional variables to compute (in addition to materializers)
+        :param overrides: Overrides to pass to execution. Optional.
+        :param inputs: Inputs to pass to execution. Optional.
+        """
```
Comment (skrawcz) [inline hamilton/driver.py:1098]: ditto - add :raises: bit

File: hamilton/io/materialization.py [hunk 3]
```diff
@@ -106,6 +106,17 @@ def _process_kwargs(
...
+        for name in self.dependencies:
+            if name in fn_graph.nodes:
+                out.append(fn_graph.nodes[name])
+            else:
+                missing_nodes.append(name)
+        if missing_nodes:
+            raise ValueError(
+                f"Materializer {self.id} has dependencies that are not in the graph: {missing_nodes}."
+            )
         return [fn_graph.nodes[name] for name in self.dependencies]
```
Comment (skrawcz) [inline hamilton/io/materialization.py:120]: i think we should add a test for this?

Comment (elijahbenizzy) [reply hamilton/io/materialization.py:120]: Its already tested in the other tests that we've added -- I think this is OK for now

Comment (skrawcz) [reply hamilton/io/materialization.py:120]: Then how did you catch it/know to add this code? 🤔

**Review by skrawcz — APPROVED** [review]

**Pull request merged**